generic: add STMicroelectronics LED1202 LED controller

innovara opened the pull request:
The Linksys MX4200v2 router doesn't have the same LED controller as the MX4200v1 or MX4300.  It comes with the STMicroelectronics LED1202.

This LED controller has a driver under development which is currently being reviewed by the respective kernel maintainers. They are currently on v11 which is included in this PR.

Additionally to the driver patches and kmod configuration, the PR makes the necessary changes to enable the driver on the MX4200v2 and amends the configuration of other devices affected by this change as the LED controller NXP PCA9633 is no common to all of them as it was originally thought.

innovara committed:
generic: add STMicroelectronics LED1202 driver

This LED controller has a driver under development which is currently being reviewed by the respective kernel maintainers. They are currently on v11 which is included here.

The LED1202 is a 12-channel low quiescent current LED driver with:
  * Supply range from 2.6 V to 5 V
  * 20 mA current capability per channel
  * 1.8 V compatible I2C control interface
  * 8-bit analog dimming individual control
  * 12-bit local PWM resolution
  * 8 programmable patterns

If the led node is present in the controller then the channel is
set to active.

The output current can be adjusted separately for each channel by 8-bit
analog (current sink input) and 12-bit digital (PWM) dimming control. The
LED1202 implements 12 low-side current generators with independent dimming
control.

Internal volatile memory allows the user to store up to 8 different patterns,
each pattern is a particular output configuration in terms of PWM
duty-cycle (on 4096 steps). Analog dimming (on 256 steps) is per channel but
common to all patterns. Each device tree LED node will have a corresponding
entry in /sys/class/leds with the label name. The brightness property
corresponds to the per channel analog dimming, while the patterns[1-8] to the
PWM dimming control.

Signed-off-by: Manuel Fombuena <fombuena@outlook.com>
Link: https://github.com/openwrt/openwrt/pull/17451
Signed-off-by: Robert Marko <robimarko@gmail.com>

robimarko commented:
Can you please split up the changes into 2 commits at least, one adding the LED controller kmod and other one actually making the qualcommax changes.

Also, you have to put the LED controller patches into generic and not qualcommax

innovara commented:
> Also, you have to put the LED controller patches into generic and not qualcommax

Given the driver is only under review, would it be `target/linux/generic/pending-6.6/8xx` the correct path/category?

robimarko commented:
Yes, pending is the right place

innovara committed:
qualcommax: ipq807x: mx4200v2:  fix LED controller

The Linksys MX4200v2 doesn't have the same LED controller as the MX4200v1 or MX4300.  It comes with the STMicroelectronics LED1202 while the others come with the NXP PCA9633.

This LED controller has a driver under development which is currently being reviewed by the respective kernel maintainers. They are currently on v11.

Apart from the changes needed on the MX4200v2, this commit also amends the configuration of other devices affected by this change as the LED controller is no common to all of them as it was originally thought.

Signed-off-by: Manuel Fombuena <fombuena@outlook.com>
Link: https://github.com/openwrt/openwrt/pull/17451
Signed-off-by: Robert Marko <robimarko@gmail.com>

robimarko commented:
You can merge the commit adding the patches to one making the kmod package

robimarko commented:
Thanks! Rebased on top of main and merged!

innovara commented:
Please note that I was still working on this PR and it was truly a draft. Maybe I should have added WIP to the subject. Unfortunately, the driver uses the `pattern` trigger only but not in the way that OpenWrt would be currently able to use it. You can't write values to `/sys/class/leds/${led}:status/pattern`. You've got to write them to `/sys/class/leds/${led}:status/hw_pattern` and then a repeat value to `/sys/class/leds/${led}:status/repeat`. You can see how you use it in leds_st1202.sh [here](https://github.com/innovara/leds_st1202). Plus the aliases `led-boot`, `led-running` and so on in `ipq8174-mx4x00.dtsi`.

As it currently is, it's a regression because the driver switches off the led and you can't switch on with OpenWrt as it is now.

robimarko commented:
Well, I was under the impression that this worked, and that draft was only until the packaging was finalized.

Do you want me to revert the device change now, or you simply want to do a follow-up?

innovara commented:
Revert for the time being please. It might take a while until I work out the rest, including patches to the driver. I am communicating with the developer so some changes might come directly from 'upstream'.

robimarko commented:
Ok, I will leave the driver backport though since it wont break anything.

innovara commented:
Great, thanks. Sorry for the misunderstanding. I was actually writing a follow up to see if you have any suggestions about how OpenWrt could use such driver.

robimarko commented:
I guess you could add a new helper that would echo the required stuff to sysfs as the default one wont cut it

innovara commented:
I thought about that, but the problem I see is that each controller using `hw_pattern` defines its own syntax. I guess that could be defined as well per device in OpenWrt. I will investigate. Thanks for your help, it's much appreciated.

robimarko commented:
But, wouldnt something like adding a new ucidef helper be perfect for this?
So you can provide the `hw_pattern` per LED?

innovara commented:
I've opened another PR to make a change to kmod-leds-st1202. That should be merged regardless the rest of the work that I'll be sending subsequently which will likely require a more in-depth review.
